rustPlatform.buildRustPackage: support direct use of Cargo.lock

danieldk opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

This change introduces the `cargoLock` argument to `buildRustPackage`,
which can be used in place of `cargo{Sha256,Hash}` or `cargoVendorDir`. It
uses the `importCargoLock` function to build the vendor
directory. Differences compared to `cargo{Sha256,Hash}`:
    
- Requires a `Cargo.lock` file.
- Does not require a Cargo hash.
- Retrieves all dependencies as fixed-output derivations.
    
This makes `buildRustPackage` much easier to use as part of a Rust
project, since it does not require updating `cargo{Sha256,Hash}` for
every change to the lock file.

To introduce this functionality, the `importCargoLock` function was
also added. This function can be used to create an output path that is
a cargo vendor directory. In contrast to e.g. `fetchCargoTarball` all
the dependent crates are fetched using fixed-output derivations. The
hashes for the fixed-output derivations are gathered from the
`Cargo.lock` file.
    
Usage is very simple, e.g.:

```
importCargoLock {
  lockFile = ./Cargo.lock;
}
```
    
would use the lockfile from the current directory.
    
The implementation of this function is based on Eelco Dolstra's
import-cargo:
    
https://github.com/edolstra/import-cargo/blob/master/flake.nix
    
Compared to `cargo-import`:

- We use `fetchgit` in place of `builtins.fetchGit`. This requires specifying output hashes, which can be done through the `outputHash` attribute (see the documentation). Since crates.io does not allow uploads with non-crates.io dependencies, git dependencies do not occur much in practive.
- Sync to current cargo vendoring.

cc @Ma27, @edolstra 

Related issue: #89563

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

danieldk committed:
rustPlatform.importCargoLock: init

This function can be used to create an output path that is a cargo
vendor directory. In contrast to e.g. fetchCargoTarball all the
dependent crates are fetched using fixed-output derivations. The
hashes for the fixed-output derivations are gathered from the
Cargo.lock file.

Usage is very simple, e.g.:

importCargoLock {
  lockFile = ./Cargo.lock;
}

would use the lockfile from the current directory.

The implementation of this function is based on Eelco Dolstra's
import-cargo:

https://github.com/edolstra/import-cargo/blob/master/flake.nix

Compared to upstream:

- We use fetchgit in place of builtins.fetchGit.
- Sync to current cargo vendoring.

danieldk committed:
rustPlatform.buildRustPackage: support direct use of Cargo.lock

This change introduces the cargoLock argument to buildRustPackage,
which can be used in place of cargo{Sha256,Hash} or cargoVendorDir. It
uses the importCargoLock function to build the vendor
directory. Differences compared to cargo{Sha256,Hash}:

- Requires a Cargo.lock file.
- Does not require a Cargo hash.
- Retrieves all dependencies as fixed-output derivations.

This makes buildRustPackage much easier to use as part of a Rust
project, since it does not require updating cargo{Sha256,Hash} for
every change to the lock file.

andir commented on pkgs/build-support/rust/import-cargo-lock.nix:
```diff
@@ -0,0 +1,104 @@
+{ fetchgit, fetchurl, lib, runCommand }:
```
While I like this approach it is a bit of an issue with cargo workspaces that contain more than one crate. I often just hacki in an override mechanism to select output dirs based on a crate name an maintain that.

@alyssais did some work on that issue in a branch of theirs. Perhaps that approach might fit in here.

danieldk replied:
I think my latest changes should resolve this issue (if I understand it correctly). E.g. using the `rand` crate works, even though it is part of a workspace and uses other dependencies from the workspace (see the added tests).

andir reviewed (commented):
This is the right direction. Less abuse of FODs for fetching a random set of dependencies.

If we merge this in we need at least one test that checks that the feature isn't broken. Any small packaging test might do as long as it makes use of outputHash and whatever else we might end up with.

Mic92 commented:
I also agree with this approach but would let the review up to @andir as he has way more experience of subtile implementation details of Rust's dependency resolution.

danieldk committed:
rustPlatform.importCargoLock: add test cases for importCargoLock

danieldk commented on pkgs/build-support/rust/default.nix:
```diff
@@ -55,19 +57,21 @@
 , ... } @ args:
 
-assert cargoVendorDir == null -> !(cargoSha256 == "" && cargoHash == "");
+assert cargoVendorDir == null && cargoLock == null -> !(cargoSha256 == "" && cargoHash == "");
```
Fixed.

danieldk commented:
> If we merge this in we need at least one test that checks that the feature isn't broken. Any small packaging test might do as long as it makes use of outputHash and whatever else we might end up with.

Added three tests:

* A Rust crate with a crates.io dependency
* A Rust crate with a git dependency with no rev specified in `Cargo.toml`.
* A Rust crate with a git dependency with a rev specified in `Cargo.toml`.

I used `rand` as the dependency, since it it is part of a Cargo workspace and uses dependencies from the same workspace.

To avoid that a user has to specify the hashes for all the dependencies from the crate, I changed `importCargoLock` such that when a user specifies:

```nix
outputHash = {
  "rand-0.8.3" = "sha256-CnMJWpSnU6slmCJhbxaXq0ikxVZDk4SQwmFYNlSEQnk=";
};
```
`rand-0.8.3` is mapped to its git commit SHA from the lockfile. Each dependency that uses the same git commit SHA will then use the same output hash. I think this is more ergonomic than asking the user to use something like

```nix
outputHash = {
  "f0e01ee0a7257753cc51b291f62666f4765923ef" = "sha256-CnMJWpSnU6slmCJhbxaXq0ikxVZDk4SQwmFYNlSEQnk=";
};
```

andir commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -107,6 +107,53 @@ rustPlatform.buildRustPackage rec {
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHash = {
+      "finalfusion-0.14.0" = "sha256-BCSPsshTB2G2KV76xwU1LgovewrhF04O5FCz+5texJ0=";
```
Can we stick with to non-sri hashes for now until the hash-mismatch output in stable nix produces these?

andir commented on pkgs/build-support/rust/import-cargo-lock.nix:
```diff
@@ -0,0 +1,142 @@
+      ''
+      else if gitParts != null then
+      let
+        sha256 = gitShaOutputHash.${gitParts.sha} or lib.fakeHash;
```
Is passing the fake hash implicitly a good idea here? Maybe we should at least also log some tracing so the users gets a heads-up on why that is happening and where the hash-mismatch will come from.

danieldk replied:
Ok, it now gives output like:

```
trace: Hash for git dependency rand-0.8.3 is missing, using fake hash.
```

danieldk replied:
Maybe it should throw and give useful instructions. The traces are really easy to miss. I’ll change this later today.

danieldk replied:
Updated to throw an error. Example:

```
error: No hash was found while vendoring the git dependency rand-0.8.3. You can add
       a hash through the `outputHash` argument of `importCargoLock`:

       outputHash = {
         "rand-0.8.3" = "<hash>";
       };

       If you use `buildRustPackage`, you can add this attribute to the `cargoLock`
       attribute set.
(use '--show-trace' to show detailed location information)
```

andir reviewed (commented):
In general it looks nice. I haven't tested it yet and thus there is no +1. I left a few minor remarks.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/why-dont-nix-hashes-use-base-16/11325/27

danieldk commented:
> In general it looks nice. I haven't tested it yet and thus there is no +1. I left a few minor remarks.

Did you find the time to test this?

andir commented:
> > In general it looks nice. I haven't tested it yet and thus there is no +1. I left a few minor remarks.
> 
> Did you find the time to test this?

I just tried to convert a random package from nixpkgs to this. I picked `bitwarden_rs` and applied this patch:

```
diff --git a/pkgs/tools/security/bitwarden_rs/default.nix b/pkgs/tools/security/bitwarden_rs/default.nix
index 5676e4005c1..da18f9fa58d 100644
--- a/pkgs/tools/security/bitwarden_rs/default.nix
+++ b/pkgs/tools/security/bitwarden_rs/default.nix
@@ -17,6 +17,15 @@ in rustPlatform.buildRustPackage rec {
     sha256 = "1ncy4iwmdzdp8rv1gc5i4s1rp97d94n4l4bh08v6w4zdpx0zn8b9";
   };

+  cargoLock = {
+    lockFile = src + "/Cargo.lock";
+    outputHash = {
+      "data-url-0.1.0" = "1lx660kx4143nppjc7cyfbhmmvy4ifljvb2imcqdxkqbqfaj7sbq";
+      "devise-0.3.0" = "0li9n5pviky3i4sb4h7p78vydgbx1nl3hrlynf42n5mvl8wc9fbj";
+      "rocket-0.5.0-dev" = "19lyljqhi6xszg86zl6wzd14bnjdb56dkcgqg2s3qz7bs8dgm4mi";
+    };
+  };
+
   nativeBuildInputs = [ pkg-config ];
   buildInputs = with lib; [ openssl ]
     ++ optionals stdenv.isDarwin [ Security CoreServices ]
@@ -25,7 +34,7 @@ in rustPlatform.buildRustPackage rec {

   RUSTC_BOOTSTRAP = 1;

-  cargoSha256 = "139by5y2ma3v52nabzr5man1qy395rchs2dlivkj9xi829kg4mcr";
+  # cargoSha256 = "139by5y2ma3v52nabzr5man1qy395rchs2dlivkj9xi829kg4mcr";
   cargoBuildFlags = [ featuresFlag ];

   checkPhase = ''
```

but it fails like this:

```
building '/nix/store/mhqcy8lx04ikyhvs4sx3y8mwpb61fk29-bitwarden_rs-1.20.0.drv'...
unpacking sources
unpacking source archive /nix/store/a02x61417w9yq2c3lf9c995azpk0ck22-source
source root is source
Executing cargoSetupPostUnpackHook
unpacking source archive /nix/store/nlhb6r908hj0skrhifqd4wk51l0n3gb8-cargo-vendor-dir
Finished cargoSetupPostUnpackHook
patching sources
Executing cargoSetupPostPatchHook
Validating consistency between /build/source//Cargo.lock and /build/cargo-vendor-dir/Cargo.lock
Finished cargoSetupPostPatchHook
configuring
building
Executing cargoBuildHook
++ env CC_x86_64-unknown-linux-gnu=/nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/c++ CC_x86_64-unknown-linux-gnu=/nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/cc CXX_x86_64-unknown-linux-gnu=/nix/store/zzvq5qwlm2xikawfqxb0q8gl2bw391a9-gcc-wrapper-10.2.0/bin/c++ cargo build -j 24 --target x86_64-unknown-linux-gnu --frozen --release --features sqlite
error: failed to resolve patches for `https://github.com/rust-lang/crates.io-index`

Caused by:
  failed to load source for dependency `data-url`

Caused by:
  Unable to update https://github.com/servo/rust-url?rev=540ede02d0771824c0c80ff9f57fe8eff38b1291#540ede02

Caused by:
  failed to update replaced source https://github.com/servo/rust-url?rev=540ede02d0771824c0c80ff9f57fe8eff38b1291#540ede02

Caused by:
  found a virtual manifest at `/build/cargo-vendor-dir/devise_codegen-0.3.0/Cargo.toml` instead of a package manifest
```

Should it be this way?

danieldk commented:
> Should it be this way?

Looks good. The issue is some workspaces where crates are not in first-level directory structures. I have it building now, but I need to clean up the changes.

SuperSandro2000 commented on pkgs/build-support/rust/default.nix:
```diff
@@ -55,19 +57,22 @@
 
-assert cargoVendorDir == null -> !(cargoSha256 == "" && cargoHash == "");
+assert cargoVendorDir == null && cargoLock == null -> cargoSha256 == "" && cargoHash == ""
+  -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock should be set";
```
```suggestion
  -> throw "cargoSha256, cargoHash, cargoVendorDir, or cargoLock must be set";
```

danieldk replied:
Fixed.

danieldk commented:
> > Should it be this way?
> 
> Looks good. The issue is in workspaces where crates are not top-level directories. I have it building now, but I need to clean up the changes.

@andir this should work now. For crates from workspaces `cargo metadata` is now used to find the correct path of the crate within the workspace.

andir commented:
> > > Should it be this way?
> > 
> > 
> > Looks good. The issue is in workspaces where crates are not top-level directories. I have it building now, but I need to clean up the changes.
> 
> @andir this should work now. For crates from workspaces `cargo metadata` is now used to find the correct path of the crate within the workspace.

Thanks! Looks better. I'll try a few more projects to see if I can find any more obvious errors.

danieldk commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -107,6 +107,53 @@ rustPlatform.buildRustPackage rec {
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHash = {
+      "finalfusion-0.14.0" = "17f4bsdzpcshwh74w5z119xjy2if6l2wgyjy56v621skr2r8y904";
```
Done. I now throw when the git commit SHA -> hash mapping is set up and the `name-version` pair is not a git dependency. However, I had to use `deepSeq` to force evaluation for when there are no git dependencies. If you can think of a more elegant hammer, I'd love to know.

andir commented:
This looks good to me. I was able to build a rather large internal workspace using this method. I still want to look at the implementation in more detail but I can't do that tonight. Likely tomorrow. I also requested a review from @kolloch as he is usually also very knowledgeable in the area of cargo metadata.

danieldk commented:
> This looks good to me. I was able to build a rather large internal workspace using this method.

Cool! And thanks for all the comments so far!

alyssais commented:
You might want to look at my own attempt at this:

https://github.com/alyssais/nixpkgs/blob/import-cargo/pkgs/build-support/rust/import-cargo.nix

With my version, I was able to convert every single package in Nixpkgs (as of December) to it.  But to get around Cargo's annoying handling of workspaces, I had to patch Cargo (https://github.com/rust-lang/cargo/pull/9024), and upstream didn't seem interested, which soured me on the whole thing.

One thing I did that I think you should do here is make an importCargo function that returns a setupHook, which would make it possible to use outside of buildRustPackage (important for packages that are written in multiple languages).

danieldk commented:
> You might want to look at my own attempt at this:
> 
> https://github.com/alyssais/nixpkgs/blob/import-cargo/pkgs/build-support/rust/import-cargo.nix
> 
> With my version, I was able to convert every single package in Nixpkgs (as of December) to it. But to get around Cargo's annoying handling of workspaces, I had to patch Cargo ([rust-lang/cargo#9024](https://github.com/rust-lang/cargo/pull/9024)), and upstream didn't seem interested, which soured me on the whole thing.

Thanks for the pointers! I’ll try to have a look tomorrow.

> One thing I did that I think you should do here is make an importCargo function that returns a setupHook, which would make it possible to use outside of buildRustPackage (important for packages that are written in multiple languages).

I think that this should already work with the hookification of `buildRustPackage` that I did earlier this year. `cargo-setup-hook’ picks up the `cargoDeps` in multi-language derivations and `importCargoLock` should work as a drop-in replacement for `fetchCargoTarball`.  But I’ll test this tomorrow to ensure that it works properly.

danieldk commented:
> I think that this should already work with the hookification of `buildRustPackage` that I did earlier this year. `cargo-setup-hook` picks up the `cargoDeps` in multi-language derivations and `importCargoLock` should work as a drop-in replacement for `fetchCargoTarball`. But I’ll test this tomorrow to ensure that it works properly.

Works with the hooks as expected. I have added a Rust/Python project using `buildPythonPackage` as an additional test case:

https://github.com/NixOS/nixpkgs/blob/ff470c3a09fe95673108c9d7e3ad31a12d61d5ca/pkgs/build-support/rust/test/import-cargo-lock/maturin/default.nix#L20

danieldk commented:
> > I think that this should already work with the hookification of `buildRustPackage` that I did earlier this year. `cargo-setup-hook` picks up the `cargoDeps` in multi-language derivations and `importCargoLock` should work as a drop-in replacement for `fetchCargoTarball`. But I’ll test this tomorrow to ensure that it works properly.
> 
> Works with the hooks as expected. I have added a Rust/Python project using `buildPythonPackage` as an additional test case:
> 
> https://github.com/NixOS/nixpkgs/blob/ff470c3a09fe95673108c9d7e3ad31a12d61d5ca/pkgs/build-support/rust/test/import-cargo-lock/maturin/default.nix#L20

Added this to the documentation now as well.

andir reviewed (approved):
From my point of view this looks good to go in. I was about to mention the setup hook but that has already been taken care of :+1:.

If @alyssais is happy with the setup hook then lets merge this!

danieldk commented:
@alyssais any objections against merging this?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/cant-get-rustplatform-to-work-with-src/13210/3

ckauhaus commented:
This feature would be a significant step forward in packaging usability.

danieldk commented:
I would hate for this to get stuck, so unless there are objections, I'll merge this at Friday 28, 12:00 CET.

asymmetric commented on doc/languages-frameworks/rust.section.md:
```diff
@@ -107,6 +107,53 @@ rustPlatform.buildRustPackage rec {
+
+  cargoLock = {
+    lockFile = ./Cargo.lock;
+    outputHash = {
```
Should this be called `outputHashes`?

danieldk replied:
I thought a lot about this and couldn't really make up my mind. Changed to `outputHashes`.

asymmetric commented on pkgs/build-support/rust/import-cargo-lock.nix:
```diff
@@ -0,0 +1,167 @@
+
+  nameGitSha = pkg: let gitParts = parseGit pkg.source; in {
+    name = "${pkg.name}-${pkg.version}";
+    value = gitParts.sha;
```
This is where I think `gitParts` could be `null`.

danieldk replied:
Note that `nameGitSha` is only applied to git sources the `map` above.

asymmetric commented on pkgs/build-support/rust/import-cargo-lock.nix:
```diff
@@ -0,0 +1,167 @@
+      parts = builtins.match ''git\+([^?]+)(\?rev=(.*))?#(.*)?'' src;
+      rev = builtins.elemAt parts 2;
+    in
+      if parts == null then null
```
IIUC, this could mean `parseGit foo` evaluates to `null`, right? We use it below as `gitParts.sha` so wouldn't this crash at runtime?

asymmetric reviewed (commented):
This would be great to have. Left a couple of (mostly stylistic) comments.

danieldk commented on doc/languages-frameworks/rust.section.md:
````diff
@@ -107,6 +107,53 @@ rustPlatform.buildRustPackage rec {
+```
+
+The output hash can be obtained by initially leaving the hash out and building
+the derivation. Nix will then report the expected hash.
````
I copied the wording from the later section on integrating importCargoLock in mixed-language projects.

danieldk commented:
@ofborg build tests.importCargoLock

danieldk commented:
Thanks for all the reviews!

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nvfetcher-generate-nix-sources-expr-for-the-latest-version-of-packages/13530/3

bergkvist commented:
A pattern I find myself using when some upstream project doesn't have a `Cargo.lock`, and I have to provide my own:
```nix
# ...
pkgs.rustPlatform.buildRustPackage {
  # ...
  cargoLock.lockFile = ./Cargo.lock;
  cargoPatches = [
    (pkgs.runCommand "Cargo.lock.patch" { buildInputs = [ pkgs.git ]; } ''
      cp ${./Cargo.lock} Cargo.lock
      git init && git add Cargo.lock
      git diff --cached > $out
    '')
  ];
}
```

That way I don't need to manually provide the Cargo.lock patch. Maybe this patch generation could be automated in nixpkgs when cargoLock.lockFile is specified?

If I omit the cargoPatches entry, I get an error message during the build:

```
Executing cargoSetupPostPatchHook
Validating consistency between /build/source/Cargo.lock and /build/cargo-vendor-dir/Cargo.lock
/nix/store/49y3r0gr9m6k20d91kl6dgp4b9a6m72v-diffutils-3.8/bin/diff: /build/source/Cargo.lock: No such file or directory
ERROR: Missing Cargo.lock from src. Expected to find it at: /build/source/Cargo.lock
Hint: You can use the cargoPatches attribute to add a Cargo.lock manually to the build.
```

Mic92 commented:
Probably it shouldn't be even apply a patch but just copy the whole thing over?
